File: intelligence/anomaly.py
"""Unusual machine usage: deterministic rules first, Isolation Forest second.

Layer 1 catches the patterns the brief names outright (excessive idling,
seatbelt violations, overload, harsh operation, overheating).  Layer 2 catches
what the rules miss, by learning what normal looks like per machine.

C detects; B explains.  `evidence` is what B's LLM needs to write the sentence.
"""

from __future__ import annotations

import json
from datetime import datetime, timedelta, timezone
from pathlib import Path

import numpy as np
import pandas as pd

from simulator.config import DIESEL_PRICE_INR, SPECS

DATA = Path(__file__).resolve().parent.parent / "data"
MODEL_DIR = Path(__file__).resolve().parent / "models"
MODEL_FILE = MODEL_DIR / "anomaly_iforest.joblib"

# The brief defines its patterns over a 2-hour window ("idle > 45 min in a
# 2-hour window"), and its own sample rows are 2-hour aggregates.  A 30-minute
# window sits entirely inside a lunch break and flags it as excessive idling,
# so the whole layer works at the same 2-hour scale the patterns are stated in.
WINDOW_MIN = 120

# thresholds straight from section 9 of the brief
IDLE_RATIO_HIGH = 0.45
CYCLES_PER_HOUR_LOW = 30.0
OVERLOAD_RATIO = 1.10
HARSH_MULTIPLIER = 3.0
TEMP_HIGH_C = 95.0
SEATBELT_OFF_RATIO = 0.30

FEATURES = [
    "idle_ratio", "fuel_per_cycle", "cycles_per_hour", "mean_payload",
    "harsh_swing_rate", "mean_hydraulic_temp", "seatbelt_off_ratio",
]

_forest = None
_baselines: dict | None = None


# --------------------------------------------------------------------------
# window features
# --------------------------------------------------------------------------

def window_features(rows: pd.DataFrame) -> dict:
    """Collapse a window of per-minute telemetry into the model's features."""
    minutes = max(len(rows), 1)
    cycles = float(rows["load_cycles"].sum())
    fuel = float(rows["fuel_used_l"].sum())
    idle = float(rows["idle_min"].sum())
    belt_off = float((rows["seatbelt"].astype(str).str.lower() == "unfastened").sum())
    harsh = float(rows["harsh_swing"].astype(bool).sum()) if "harsh_swing" in rows else 0.0
    return {
        "idle_ratio": idle / minutes,
        "fuel_per_cycle": fuel / cycles if cycles > 0 else fuel * 10.0,
        "cycles_per_hour": cycles / minutes * 60.0,
        "mean_payload": float(rows["payload_kg"].mean()) if "payload_kg" in rows else 0.0,
        "harsh_swing_rate": harsh / minutes,
        "mean_hydraulic_temp": float(rows["hydraulic_temp_c"].mean())
        if "hydraulic_temp_c" in rows else 0.0,
        "seatbelt_off_ratio": belt_off / minutes,
        # carried along for evidence, not used as model features
        "_idle_min": idle,
        "_load_cycles": int(cycles),
        "_seatbelt_off_min": belt_off,
        "_fuel_l": fuel,
        "_minutes": minutes,
    }


def _rules(feat: dict, max_payload_kg: float, harsh_baseline: float) -> list[str]:
    """Which named patterns fired in this window."""
    fired = []
    # idling most of the window while barely producing anything
    if feat["idle_ratio"] > IDLE_RATIO_HIGH and feat["cycles_per_hour"] < CYCLES_PER_HOUR_LOW:
        fired.append("excessive_idling")
    if feat["seatbelt_off_ratio"] > SEATBELT_OFF_RATIO:
        fired.append("seatbelt_violation")
    if max_payload_kg and feat["mean_payload"] > max_payload_kg * OVERLOAD_RATIO:
        fired.append("overload")
    if harsh_baseline > 0 and feat["harsh_swing_rate"] > harsh_baseline * HARSH_MULTIPLIER:
        fired.append("harsh_operation")
    if feat["mean_hydraulic_temp"] > TEMP_HIGH_C:
        fired.append("temperature_anomaly")
    return fired


def _fuel_wasted_l(feat: dict, model: str) -> float:
    """Diesel burned while idling - the number the owner screen cares about."""
    spec = SPECS.get(model)
    idle_rate = spec.idle_fuel_lph if spec else 4.0
    return round(feat["_idle_min"] / 60.0 * idle_rate, 1)


# --------------------------------------------------------------------------
# model layer
# --------------------------------------------------------------------------

def _load_forest():
    global _forest, _baselines
    if _forest is not None:
        return _forest, _baselines
    if not MODEL_FILE.exists():
        return None, None
    import joblib

    bundle = joblib.load(MODEL_FILE)
    _forest = bundle["model"]
    _baselines = bundle.get("baselines", {})
    return _forest, _baselines


def _score(feat: dict) -> float:
    """Isolation Forest score, normalised to 0-1 (1 = most anomalous)."""
    forest, _ = _load_forest()
    if forest is None:
        return 0.0
    x = np.array([[feat[f] for f in FEATURES]], dtype=float)
    raw = float(forest.score_samples(x)[0])
    # score_samples is roughly -0.75 (very anomalous) .. -0.35 (normal)
    return float(np.clip((-raw - 0.35) / 0.35, 0.0, 1.0))


def _largest_deviation(feat: dict, baseline: dict | None) -> str:
    if not baseline:
        return "unusual_pattern"
    worst, worst_z = "unusual_pattern", 0.0
    for f in FEATURES:
        mean = baseline.get(f, {}).get("mean")
        std = baseline.get(f, {}).get("std") or 1e-6
        if mean is None:
            continue
        z = abs((feat[f] - mean) / std)
        if z > worst_z:
            worst, worst_z = f, z
    return {
        "idle_ratio": "excessive_idling",
        "seatbelt_off_ratio": "seatbelt_violation",
        "mean_payload": "overload",
        "harsh_swing_rate": "harsh_operation",
        "mean_hydraulic_temp": "temperature_anomaly",
        "fuel_per_cycle": "low_productivity",
        "cycles_per_hour": "low_productivity",
    }.get(worst, "unusual_pattern")


# --------------------------------------------------------------------------
# public API
# --------------------------------------------------------------------------

def _iso(ts) -> str:
    return pd.Timestamp(ts).strftime("%Y-%m-%dT%H:%M:%SZ")


def analyse_windows(telemetry: pd.DataFrame, window_min: int = WINDOW_MIN) -> list[dict]:
    """Score every window of a telemetry frame.  Shared by history and live."""
    _, baselines = _load_forest()
    machines = pd.read_csv(DATA / "machines.csv") if (DATA / "machines.csv").exists() else None
    model_by_machine = (
        dict(zip(machines["machine_id"], machines["model"])) if machines is not None else {}
    )

    out: list[dict] = []
    telemetry = telemetry.copy()
    telemetry["timestamp"] = pd.to_datetime(telemetry["timestamp"])
    telemetry["_window"] = telemetry["timestamp"].dt.floor(f"{window_min}min")

    counter = 0
    for (machine_id, window), rows in telemetry.groupby(["machine_id", "_window"]):
        if len(rows) < window_min // 2:
            continue
        feat = window_features(rows)
        model = model_by_machine.get(machine_id, "320")
        spec = SPECS.get(model)
        baseline = (baselines or {}).get(machine_id, {})
        harsh_baseline = baseline.get("harsh_swing_rate", {}).get("mean", 0.06)

        fired = _rules(feat, spec.max_payload_kg if spec else 0.0, harsh_baseline)
        score = _score(feat)
        if not fired and score < 0.55:
            continue

        counter += 1
        primary = fired[0] if fired else _largest_deviation(feat, baseline)
        wasted = _fuel_wasted_l(feat, model)
        out.append({
            "anomaly_id": f"AN-{counter:04d}",
            "machine_id": machine_id,
            "operator_id": str(rows["operator_id"].iloc[0]),
            "type": primary,
            "related": [f for f in fired[1:]] + (
                ["low_productivity"] if feat["cycles_per_hour"] < 30 else []
            ),
            "score": round(max(score, 0.75 if fired else score), 2),
            "detected_by": "rules" if fired else "isolation_forest",
            "window": {"start": _iso(window),
                       "end": _iso(window + timedelta(minutes=window_min))},
            "evidence": {
                "idle_min": round(feat["_idle_min"], 1),
                "load_cycles": feat["_load_cycles"],
                "seatbelt_off_min": round(feat["_seatbelt_off_min"], 1),
                "mean_payload_kg": round(feat["mean_payload"], 0),
                "mean_hydraulic_temp_c": round(feat["mean_hydraulic_temp"], 1),
                "baseline_idle_min": round(
                    baseline.get("idle_ratio", {}).get("mean", 0.2) * window_min, 1
                ),
            },
            "fuel_wasted_l": wasted,
            "fuel_cost_inr": round(wasted * DIESEL_PRICE_INR),
        })
    return out


def get_anomalies(machine_id: str | None = None, since_hours: int = 24) -> list[dict]:
    """Anomalies from the historical record, most anomalous first."""
    path = DATA / "telemetry.csv"
    if not path.exists():
        return []
    telemetry = pd.read_csv(path, parse_dates=["timestamp"])
    if machine_id:
        telemetry = telemetry[telemetry["machine_id"] == machine_id]
    if telemetry.empty:
        return []

    latest = telemetry["timestamp"].max()
    cutoff = latest - timedelta(hours=since_hours)
    recent = telemetry[telemetry["timestamp"] >= cutoff]
    if recent.empty:
        recent = telemetry

    found = analyse_windows(recent)
    found.sort(key=lambda a: (-a["score"], a["window"]["start"]))
    return found


def score_live_window(states: list[dict]) -> dict | None:
    """Score one machine's last ~60 seconds of live `machine_state` messages.

    The simulator calls this once a minute per machine.  Live states carry
    cumulative counters, so they are differenced into per-window totals first.
    """
    if len(states) < 10:
        return None

    first, last = states[0], states[-1]
    minutes = max(len(states) / 60.0, 1 / 60)
    idle_min = max(0.0, float(last["idle_min"]) - float(first["idle_min"]))
    cycles = max(0, int(last["load_cycles"]) - int(first["load_cycles"]))
    fuel = max(0.0, float(last["fuel_used_l"]) - float(first["fuel_used_l"]))
    belt_off = sum(1 for s in states if s["seatbelt"] == "unfastened")

    feat = {
        "idle_ratio": min(idle_min / minutes, 1.0),
        "fuel_per_cycle": fuel / cycles if cycles else fuel * 10.0,
        "cycles_per_hour": cycles / minutes * 60.0,
        "mean_payload": float(np.mean([s["payload_kg"] for s in states])),
        "harsh_swing_rate": 0.0,
        "mean_hydraulic_temp": float(np.mean([s["hydraulic_temp_c"] for s in states])),
        "seatbelt_off_ratio": belt_off / len(states),
        "_idle_min": idle_min,
        "_load_cycles": cycles,
        "_seatbelt_off_min": belt_off / 60.0,
        "_fuel_l": fuel,
        "_minutes": minutes,
    }

    model = str(last.get("model", "320"))
    spec = SPECS.get(model)
    fired = _rules(feat, spec.max_payload_kg if spec else 0.0, 0.06)
    score = _score(feat)
    if not fired and score < 0.6:
        return None

    wasted = _fuel_wasted_l(feat, model)
    return {
        "machine_id": last["machine_id"],
        "operator_id": last["operator_id"],
        "type": fired[0] if fired else "unusual_pattern",
        "related": fired[1:],
        "score": round(max(score, 0.75 if fired else score), 2),
        "detected_by": "rules" if fired else "isolation_forest",
        "evidence": {
            "idle_min": round(idle_min, 1),
            "load_cycles": cycles,
            "seatbelt_off_s": belt_off,
        },
        "fuel_wasted_l": wasted,
        "fuel_cost_inr": round(wasted * DIESEL_PRICE_INR),
    }


def score_brief_sample() -> list[dict]:
    """Score the exact four rows from the brief.

    They are 2-hour aggregate windows, so they are scored as pre-aggregated
    windows rather than re-derived from per-minute data.  Used on the pitch
    slide: "our model flags the rows you gave us."
    """
    path = DATA / "brief_sample.csv"
    if not path.exists():
        return []
    rows = pd.read_csv(path)
    out = []
    for _, r in rows.iterrows():
        minutes = 120.0
        feat = {
            "idle_ratio": float(r["idling_time_min"]) / minutes,
            "fuel_per_cycle": float(r["fuel_used_l"]) / max(int(r["load_cycles"]), 1),
            "cycles_per_hour": int(r["load_cycles"]) / minutes * 60.0,
            "mean_payload": 1_500.0,
            "harsh_swing_rate": 0.0,
            "mean_hydraulic_temp": 78.0,
            "seatbelt_off_ratio": (
                float(r["idling_time_min"]) / minutes
                if str(r["seatbelt_status"]).lower() == "unfastened" else 0.0
            ),
            "_idle_min": float(r["idling_time_min"]),
            "_load_cycles": int(r["load_cycles"]),
            "_seatbelt_off_min": (
                float(r["idling_time_min"])
                if str(r["seatbelt_status"]).lower() == "unfastened" else 0.0
            ),
            "_fuel_l": float(r["fuel_used_l"]),
            "_minutes": minutes,
        }
        fired = _rules(feat, SPECS["320"].max_payload_kg, 0.06)
        out.append({
            "timestamp": str(r["timestamp"]),
            "machine_id": str(r["machine_id"]),
            "expected_alert": str(r["safety_alert_triggered"]) == "Yes",
            "flagged": bool(fired),
            "types": fired,
            "score": round(_score(feat), 2),
            "evidence": {
                "idle_min": float(r["idling_time_min"]),
                "load_cycles": int(r["load_cycles"]),
                "seatbelt": str(r["seatbelt_status"]),
            },
        })
    return out

File: src/lib/intel/anomaly.ts
/**
 * Unusual machine usage, running in the browser.
 *
 * Two layers, mirroring `intelligence/anomaly.py`:
 *
 *  1. Deterministic rules for the patterns the brief names outright — excessive
 *     idling, seatbelt violations, overload, harsh operation, overheating. The
 *     thresholds are the ones in the brief, exported rather than retyped.
 *  2. What the rules miss, caught by distance from what this specific machine
 *     normally does. The Isolation Forest itself cannot run here, but the
 *     per-machine baselines it was fitted alongside are means and standard
 *     deviations, so this layer scores z-distance against them. Same inputs,
 *     same features, a blunter decision boundary.
 *
 * "Cycles near zero" is deliberately relative: a 745 truck does ~21 loads an
 * hour where a 320 excavator does ~150 cycles, so any absolute floor either
 * flags every truck or misses every excavator.
 */

import { intel, specFor, dieselCostInr } from "./model";

const T = intel.anomaly.thresholds;
const BASELINES = intel.anomaly.baselines;

export type AnomalyType =
  | "excessive_idling"
  | "seatbelt_violation"
  | "overload"
  | "harsh_operation"
  | "temperature_anomaly"
  | "low_productivity"
  | "unusual_pattern";

/** One sample of live machine state, with counters as the machine reports them. */
export interface TelemetrySample {
  t: number;
  /** Cumulative minutes idled this shift. */
  idleMin: number;
  /** Cumulative load cycles this shift. */
  loadCycles: number;
  /** Cumulative litres burnt this shift. */
  fuelUsedL: number;
  payloadKg: number;
  hydraulicTemperatureC: number;
  seatbeltOff: boolean;
  harshSwing?: boolean;
}

export interface WindowFeatures {
  idleRatio: number;
  fuelPerCycle: number;
  cyclesPerHour: number;
  meanPayload: number;
  peakPayload: number;
  harshSwingRate: number;
  meanHydraulicTemp: number;
  peakHydraulicTemp: number;
  seatbeltOffRatio: number;
  /* carried for evidence, not scored */
  idleMinutes: number;
  cycles: number;
  seatbeltOffMinutes: number;
  fuelL: number;
  minutes: number;
}

export interface AnomalyFinding {
  machineId: string;
  type: AnomalyType;
  related: AnomalyType[];
  /** 0-1, 1 being most unusual. */
  score: number;
  detectedBy: "rules" | "baseline_deviation";
  /** e.g. "idle 63 min against a 24 min norm". */
  deviation: string;
  evidence: {
    idleMinutes: number;
    cycles: number;
    seatbeltOffMinutes: number;
    meanPayloadKg: number;
    peakPayloadKg: number;
    peakHydraulicTempC: number;
    baselineIdleMinutes: number;
  };
  fuelWastedL: number;
  fuelCostInr: number;
  features: WindowFeatures;
}

/** The features the baselines are keyed by, in the fitted order. */
const SCORED: (keyof WindowFeatures)[] = [
  "idleRatio",
  "fuelPerCycle",
  "cyclesPerHour",
  "meanPayload",
  "peakPayload",
  "harshSwingRate",
  "meanHydraulicTemp",
  "peakHydraulicTemp",
  "seatbeltOffRatio",
];

/** The baselines came out of Python under their snake_case names. */
const BASELINE_KEY: Record<string, string> = {
  idleRatio: "idle_ratio",
  fuelPerCycle: "fuel_per_cycle",
  cyclesPerHour: "cycles_per_hour",
  meanPayload: "mean_payload",
  peakPayload: "peak_payload",
  harshSwingRate: "harsh_swing_rate",
  meanHydraulicTemp: "mean_hydraulic_temp",
  peakHydraulicTemp: "peak_hydraulic_temp",
  seatbeltOffRatio: "seatbelt_off_ratio",
};

function quantile(values: number[], q: number): number {
  if (!values.length) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const pos = (sorted.length - 1) * q;
  const lo = Math.floor(pos);
  const hi = Math.ceil(pos);
  return sorted[lo]! + (sorted[hi]! - sorted[lo]!) * (pos - lo);
}

const mean = (values: number[]): number =>
  values.length ? values.reduce((a, b) => a + b, 0) / values.length : 0;

/**
 * Collapse a run of live samples into the model's features.
 *
 * Counters are cumulative, so they are differenced across the window first —
 * the same thing `score_live_window` does on the Python side.
 */
export function windowFeatures(samples: TelemetrySample[]): WindowFeatures {
  const first = samples[0]!;
  const last = samples[samples.length - 1]!;
  const minutes = Math.max(Math.max(last.t - first.t, 1000) / 60000, 1 / 60);

  const idleMinutes = Math.max(0, last.idleMin - first.idleMin);
  const cycles = Math.max(0, last.loadCycles - first.loadCycles);
  const fuelL = Math.max(0, last.fuelUsedL - first.fuelUsedL);
  const beltOffSamples = samples.filter((s) => s.seatbeltOff).length;
  const harsh = samples.filter((s) => s.harshSwing).length;

  const payloads = samples.map((s) => s.payloadKg);
  const temps = samples.map((s) => s.hydraulicTemperatureC);

  return {
    idleRatio: Math.min(idleMinutes / minutes, 1),
    // With no cycles at all, fuel burnt is pure waste; the x10 keeps that
    // visible to the scorer instead of dividing by zero.
    fuelPerCycle: cycles > 0 ? fuelL / cycles : fuelL * 10,
    cyclesPerHour: (cycles / minutes) * 60,
    meanPayload: mean(payloads),
    // 90th percentile rather than the outright max, so one bad sample is not an
    // overload but a sustained excursion is.
    peakPayload: quantile(payloads, 0.9),
    harshSwingRate: harsh / samples.length,
    meanHydraulicTemp: mean(temps),
    peakHydraulicTemp: quantile(temps, 0.9),
    seatbeltOffRatio: beltOffSamples / samples.length,
    idleMinutes,
    cycles,
    seatbeltOffMinutes: (beltOffSamples / samples.length) * minutes,
    fuelL,
    minutes,
  };
}

/** Which named patterns fired in this window. */
function firedRules(
  f: WindowFeatures,
  maxPayloadKg: number,
  baseline: Record<string, { mean: number; std: number }>,
): AnomalyType[] {
  const harshBaseline = baseline.harsh_swing_rate?.mean ?? 0.06;
  const normalCph = baseline.cycles_per_hour?.mean;
  const productivityFloor = normalCph
    ? normalCph * T.productivityLowRatio
    : T.cyclesPerHourFallback;

  const fired: AnomalyType[] = [];
  // Idling most of the window *and* producing well below this machine's norm.
  // Either alone is a lunch break or a slow patch; together it is waste.
  if (f.idleRatio > T.idleRatioHigh && f.cyclesPerHour < productivityFloor) {
    fired.push("excessive_idling");
  }
  if (f.seatbeltOffRatio > T.seatbeltOffRatio) fired.push("seatbelt_violation");
  if (maxPayloadKg && f.peakPayload > maxPayloadKg * T.overloadRatio) fired.push("overload");
  if (harshBaseline > 0 && f.harshSwingRate > harshBaseline * T.harshMultiplier) {
    fired.push("harsh_operation");
  }
  if (f.peakHydraulicTemp > T.tempHighC) fired.push("temperature_anomaly");
  return fired;
}

/**
 * How each rule states its own case.
 *
 * The deviation layer describes a finding by naming the feature furthest from
 * normal. A rule needs no baseline to say what it saw, and saying it from the
 * rule's own numbers keeps the sentence true when the baselines are not in play.
 */
const RULE_DEVIATION: Record<AnomalyType, (f: WindowFeatures) => string> = {
  excessive_idling: (f) =>
    `idle ${f.idleMinutes.toFixed(0)} of ${f.minutes.toFixed(0)} minutes with ${f.cycles} load cycles`,
  seatbelt_violation: (f) =>
    `belt unfastened for ${(f.seatbeltOffRatio * 100).toFixed(0)}% of the window`,
  overload: (f) => `payload peaked at ${Math.round(f.peakPayload).toLocaleString()} kg`,
  harsh_operation: (f) => `harsh swings on ${(f.harshSwingRate * 100).toFixed(0)}% of samples`,
  temperature_anomaly: (f) => `hydraulics peaked at ${f.peakHydraulicTemp.toFixed(0)} °C`,
  low_productivity: (f) => `${f.cyclesPerHour.toFixed(0)} cycles/h`,
  unusual_pattern: () => "outside this machine's normal range",
};

const DEVIATION_TYPE: Record<string, AnomalyType> = {
  idleRatio: "excessive_idling",
  seatbeltOffRatio: "seatbelt_violation",
  meanPayload: "overload",
  peakPayload: "overload",
  harshSwingRate: "harsh_operation",
  meanHydraulicTemp: "temperature_anomaly",
  peakHydraulicTemp: "temperature_anomaly",
  fuelPerCycle: "low_productivity",
  cyclesPerHour: "low_productivity",
};

/**
 * How far out of its normal range one feature is.
 *
 * The standard deviation is floored against the mean before dividing. Some
 * baselines are degenerate — a machine whose payload never varies has a std of
 * 1e-6 — and dividing by that turns a rounding difference into a z-score of
 * thousands. The floor is what keeps the second layer from crying wolf every
 * time one feature happens to be constant in the history.
 */
function zScore(value: number, stat: { mean: number; std: number }): number {
  const spread = Math.max(stat.std, Math.abs(stat.mean) * 0.05, 1e-3);
  return Math.abs((value - stat.mean) / spread);
}

/** Per-feature distances from this machine's normal, worst first. */
function deviations(
  f: WindowFeatures,
  baseline: Record<string, { mean: number; std: number }>,
): { feature: keyof WindowFeatures; z: number }[] {
  const out: { feature: keyof WindowFeatures; z: number }[] = [];
  for (const key of SCORED) {
    const stat = baseline[BASELINE_KEY[key]!];
    if (!stat) continue;
    out.push({ feature: key, z: zScore(f[key], stat) });
  }
  return out.sort((a, b) => b.z - a.z);
}

/** Diesel burnt while producing nothing — the number the owner screen cares about. */
function fuelWastedL(f: WindowFeatures, model: string): number {
  const idleRate = specFor(model)?.idleFuelLph ?? 4;
  return Math.round((f.idleMinutes / 60) * idleRate * 10) / 10;
}

function readable(
  feature: keyof WindowFeatures,
  f: WindowFeatures,
  stat: { mean: number },
): string {
  switch (feature) {
    case "idleRatio":
      return `idle ${f.idleMinutes.toFixed(0)} min against a ${(stat.mean * f.minutes).toFixed(0)} min norm`;
    case "seatbeltOffRatio":
      return `belt unfastened ${(f.seatbeltOffRatio * 100).toFixed(0)}% of the window`;
    case "cyclesPerHour":
      return `${f.cyclesPerHour.toFixed(0)} cycles/h against ${stat.mean.toFixed(0)} normally`;
    case "peakPayload":
    case "meanPayload":
      return `payload peaked at ${Math.round(f.peakPayload).toLocaleString()} kg`;
    case "peakHydraulicTemp":
    case "meanHydraulicTemp":
      return `hydraulics peaked at ${f.peakHydraulicTemp.toFixed(0)} °C`;
    case "harshSwingRate":
      return `harsh swings on ${(f.harshSwingRate * 100).toFixed(0)}% of samples`;
    case "fuelPerCycle":
      return `${f.fuelPerCycle.toFixed(1)} L per cycle against ${stat.mean.toFixed(1)} normally`;
    default:
      return "outside this machine's normal range";
  }
}

/**
 * Totals for a window that was aggregated elsewhere — a day of history, or the
 * two-hour rows the brief supplies — turned into the same features live samples
 * produce, so both go through one scorer.
 */
export function featuresFromTotals(totals: {
  minutes: number;
  idleMinutes: number;
  cycles: number;
  fuelL: number;
  seatbeltOffMinutes: number;
  meanPayload?: number;
  peakPayload?: number;
  meanHydraulicTemp?: number;
  peakHydraulicTemp?: number;
  harshSwingRate?: number;
}): WindowFeatures {
  const minutes = Math.max(totals.minutes, 1);
  return {
    idleRatio: Math.min(totals.idleMinutes / minutes, 1),
    fuelPerCycle: totals.cycles > 0 ? totals.fuelL / totals.cycles : totals.fuelL * 10,
    cyclesPerHour: (totals.cycles / minutes) * 60,
    meanPayload: totals.meanPayload ?? 0,
    peakPayload: totals.peakPayload ?? totals.meanPayload ?? 0,
    harshSwingRate: totals.harshSwingRate ?? 0,
    meanHydraulicTemp: totals.meanHydraulicTemp ?? 0,
    peakHydraulicTemp: totals.peakHydraulicTemp ?? totals.meanHydraulicTemp ?? 0,
    seatbeltOffRatio: Math.min(totals.seatbeltOffMinutes / minutes, 1),
    idleMinutes: totals.idleMinutes,
    cycles: totals.cycles,
    seatbeltOffMinutes: totals.seatbeltOffMinutes,
    fuelL: totals.fuelL,
    minutes,
  };
}

/**
 * Score one machine's recent window. Returns null when nothing is unusual —
 * which is most of the time, and is the point.
 */
export interface ScoreOptions {
  /**
   * Whether to run the second layer at all.
   *
   * The baselines describe the *historical* telemetry — what a machine's own
   * per-minute record looked like over thirty days. Scoring a live window
   * against them only works if the live feed is on the same scale, and the
   * simulated site is not: its payloads and cycle rates come from machine specs
   * rather than from the generated history, so a z-distance between the two
   * measures the gap between two data sources rather than anything about the
   * machine. So history gets both layers, and the live shift gets the rules —
   * which are absolute thresholds and need no baseline to be meaningful.
   */
  useBaselines?: boolean;
  /**
   * How much of the window has to be there before anything can fire.
   *
   * Without this a machine flagged as idling has been watched for four seconds,
   * and the evidence line reads "idle 0 of 0 minutes". The Python side refuses
   * to score a window less than half full for the same reason.
   */
  minSamples?: number;
}

export function scoreWindow(
  machineId: string,
  machineModel: string,
  samples: TelemetrySample[],
  options: ScoreOptions = {},
): AnomalyFinding | null {
  if (samples.length < (options.minSamples ?? 10)) return null;
  return scoreFeatures(machineId, machineModel, windowFeatures(samples), options);
}

/** The scorer itself, once the window is features either way. */
export function scoreFeatures(
  machineId: string,
  machineModel: string,
  f: WindowFeatures,
  options: ScoreOptions = {},
): AnomalyFinding | null {
  const useBaselines = options.useBaselines ?? true;
  const baseline = BASELINES[machineId] ?? {};
  const spec = specFor(machineModel);
  const fired = firedRules(f, spec?.maxPayloadKg ?? 0, useBaselines ? baseline : {});

  const ranked = useBaselines ? deviations(f, baseline) : [];
  const feature = ranked[0]?.feature ?? null;

  // The mean of the three worst features, not the single worst. An unusual
  // shift shows up across several of them at once; one feature alone drifting
  // is how the forest's false positives looked, and averaging suppresses them.
  const top = ranked.slice(0, 3);
  const z = top.length ? top.reduce((sum, d) => sum + d.z, 0) / top.length : 0;
  // Six deviations is full scale, putting the 0.55 cut-off a little over three.
  const deviationScore = Math.min(1, z / 6);
  if (!fired.length && (!useBaselines || deviationScore < 0.55)) return null;

  // Co-occurrence is the strong signal — the brief's own flagged rows are the
  // ones where idling and an unfastened belt happen together.
  const ruleScore = fired.length ? Math.min(0.95, 0.75 + (fired.length - 1) * 0.1) : 0;
  const score = Math.max(ruleScore, deviationScore);

  const primary: AnomalyType =
    fired[0] ?? (feature ? (DEVIATION_TYPE[feature] ?? "unusual_pattern") : "unusual_pattern");

  const related = [...new Set(fired.slice(1))].filter((r) => r !== primary);
  const normalCph = useBaselines ? baseline.cycles_per_hour?.mean : undefined;
  if (
    normalCph &&
    f.cyclesPerHour < normalCph * T.productivityLowRatio &&
    primary !== "low_productivity"
  ) {
    related.push("low_productivity");
  }

  const wasted = fuelWastedL(f, machineModel);
  const stat = feature ? baseline[BASELINE_KEY[feature]!] : undefined;
  // A rule states its own case; only a deviation finding needs the baseline to
  // explain itself.
  const deviation = fired.length
    ? RULE_DEVIATION[primary](f)
    : feature && stat
      ? readable(feature, f, stat)
      : "outside this machine's normal range";

  return {
    machineId,
    type: primary,
    related,
    score: Math.round(score * 100) / 100,
    detectedBy: fired.length ? "rules" : "baseline_deviation",
    deviation,
    evidence: {
      idleMinutes: Math.round(f.idleMinutes * 10) / 10,
      cycles: f.cycles,
      seatbeltOffMinutes: Math.round(f.seatbeltOffMinutes * 10) / 10,
      meanPayloadKg: Math.round(f.meanPayload),
      peakPayloadKg: Math.round(f.peakPayload),
      peakHydraulicTempC: Math.round(f.peakHydraulicTemp * 10) / 10,
      baselineIdleMinutes: Math.round((baseline.idle_ratio?.mean ?? 0.2) * f.minutes * 10) / 10,
    },
    fuelWastedL: wasted,
    fuelCostInr: Math.round(dieselCostInr(wasted)),
    features: f,
  };
}

export const ANOMALY_LABEL: Record<AnomalyType, string> = {
  excessive_idling: "Excessive idling",
  seatbelt_violation: "Seatbelt violation",
  overload: "Overload",
  harsh_operation: "Harsh operation",
  temperature_anomaly: "Temperature anomaly",
  low_productivity: "Low productivity",
  unusual_pattern: "Unusual pattern",
};

export const anomalyMetrics = intel.anomaly.metrics;
